src/main/kotlin/com/amazon/opendistroforelasticsearch/indexmanagement/indexstatemanagement/SkipExecution.kt

```diff
@@ -0,0 +1,77 @@
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.opendistroforelasticsearch.indexmanagement.indexstatemanagement
+
+import com.amazon.opendistroforelasticsearch.indexmanagement.util.OpenForTesting
+import org.apache.logging.log4j.LogManager
+import org.elasticsearch.action.ActionListener
+import org.elasticsearch.action.admin.cluster.node.info.NodesInfoAction
+import org.elasticsearch.action.admin.cluster.node.info.NodesInfoRequest
+import org.elasticsearch.action.admin.cluster.node.info.NodesInfoResponse
+import org.elasticsearch.action.admin.cluster.node.info.PluginsAndModules
+import org.elasticsearch.client.Client
+import org.elasticsearch.cluster.ClusterChangedEvent
+import org.elasticsearch.cluster.ClusterStateListener
+import org.elasticsearch.cluster.service.ClusterService
+
+@OpenForTesting
+class SkipExecution(
+    private val client: Client,
+    private val clusterService: ClusterService
+) : ClusterStateListener {
+    private val logger = LogManager.getLogger(javaClass)
+
+    @Volatile final var flag: Boolean = false
+        private set
+
+    init {
+        clusterService.addListener(this)
+    }
+
+    override fun clusterChanged(event: ClusterChangedEvent) {
+        if (event.nodesChanged() || event.isNewCluster) {
+            sweepISMPluginVersion()
+        }
+    }
+
+    fun sweepISMPluginVersion() {
+        // if old version ISM plugin exists (2 versions ISM in one cluster), set skip flag to true
+        val request = NodesInfoRequest().clear().addMetric("plugins")
+        client.execute(NodesInfoAction.INSTANCE, request, object : ActionListener<NodesInfoResponse> {
+            override fun onResponse(response: NodesInfoResponse) {
+                flag = false
+                val versionSet = mutableSetOf<String>()
+
+                response.nodes.map { it.getInfo(PluginsAndModules::class.java).pluginInfos }
+                    .forEach { it.forEach { nodePlugin ->
+                        if (nodePlugin.name == "opendistro-index-management" ||
+                            nodePlugin.name == "opendistro_index_management") {
+                            versionSet.add(nodePlugin.version)
+                        }
+                    } }
+
+                if (versionSet.size > 1) {
+                    flag = true
+                    logger.info("There are multiple versions of Index Management plugins in the cluster: $versionSet")
+                }
+            }
+
+            override fun onFailure(e: Exception) {
```

Comment:
What happens if flag is set to true and then this runs again and **should** set it back to false, but it fails here. Now no new nodes are added for a month and flag is stuck at true?

Reply:
I think we need to either assume flag is false here and log a warning or error saying the sweep failed and we are assuming the versions are consistent or there needs to be a self-healing and/or a way to manually set the flag to false in case it's stuck at true without having to kill a node to force this code to execute. I'm leaning towards just set the flag to false here and log out an error.. what do you think @bowenlan-amzn ?

Reply:
it's a tradeoff decision; I will take your choice :) but still give some thinking

This call is triggered when node changes, if it failes, I am thinking of treating it as a bad cluster state, and set the flag=true to stop runner running. And assuming it will have future node change event which can set flag back to false.

Reply:
But what happens when there is no future node change event? And then it's set to true and their ISM jobs stop running and rollovers don't happen and their cluster becomes unstable etc.

Reply:
I can accept this.
The only bad situation I can think of is we set this to false during upgrade, it move some metadata to config index; but the upgrade cannot succeeed and need to go back to old version, then these jobs would lost metadata

Reply:
Can you not have that problem with other way though? Like what happens when you have a job in the runner check skipFlag and it's false and then upgrade starts and flag is set to true, but now job in runner is already through that check and migrates metadata to index and then rollback happens?

Reply:
I don't think there is any rollback from this, what happens if upgrade goes through, user is unhappy with some issues and wants it reverted and then back to old plugin? All jobs would be broken then right?

Reply:
Choose to set flag = flase here when there is an exception